verifier/verifier.go

```diff
@@ -518,6 +550,76 @@ func verifyAuthenticTimestamp(outcome *notation.VerificationOutcome) *notation.V
 	}
 }
 
+func verifyRevocation(outcome *notation.VerificationOutcome, r revocation.Revocation, logger log.Logger) *notation.ValidationResult {
+	if r == nil {
+		return &notation.ValidationResult{
+			Type:   trustpolicy.TypeRevocation,
+			Action: outcome.VerificationLevel.Enforcement[trustpolicy.TypeRevocation],
+			Error:  fmt.Errorf("unable to check revocation status, revocation client cannot be nil"),
+		}
+	}
+
+	authenticSigningTime, err := outcome.EnvelopeContent.SignerInfo.AuthenticSigningTime()
+	if err != nil {
+		logger.Debugf("not using authentic signing time due to error retrieving AuthenticSigningTime, err: %v", err)
+	}
+	certResults, err := r.Validate(outcome.EnvelopeContent.SignerInfo.CertificateChain, authenticSigningTime)
+	if err != nil {
+		logger.Debug("error while checking revocation status, err: %s", err.Error())
+		return &notation.ValidationResult{
+			Type:   trustpolicy.TypeRevocation,
+			Action: outcome.VerificationLevel.Enforcement[trustpolicy.TypeRevocation],
+			Error:  fmt.Errorf("unable to check revocation status, err: %s", err.Error()),
+		}
+	}
+
+	result := &notation.ValidationResult{
+		Type:   trustpolicy.TypeRevocation,
+		Action: outcome.VerificationLevel.Enforcement[trustpolicy.TypeRevocation],
+	}
+
+	finalResult := revocationresult.ResultUnknown
+	numOKResults := 0
+	var problematicCertSubject string
+	revokedFound := false
+	var revokedCertSubject string
+	for i := len(certResults) - 1; i >= 0; i-- {
+		if certResults[i].ServerResults[0].Error != nil {
+			logger.Debugf("error for certificate #%d in chain with subject %v for server %q: %v", (i + 1), outcome.EnvelopeContent.SignerInfo.CertificateChain[i].Subject.String(), certResults[i].ServerResults[0].Server, certResults[i].ServerResults[0].Error)
+		}
+
+		if certResults[i].Result == revocationresult.ResultOK || certResults[i].Result == revocationresult.ResultNonRevokable {
+			numOKResults++
+		} else {
+			finalResult = certResults[i].Result
+			problematicCertSubject = outcome.EnvelopeContent.SignerInfo.CertificateChain[i].Subject.String()
+			if certResults[i].Result == revocationresult.ResultRevoked {
+				revokedFound = true
+				revokedCertSubject = problematicCertSubject
```

Comment:
This has been discussed a few times (most recently here: https://github.com/notaryproject/notation-go/pull/295#discussion_r1169417279)

TLDR: There are 2 problems with this:
 - Returning early means we wouldn't log every cert
 - If there is a revoked cert closer to the leaf, it wouldn't reflect that cert's subject

The current implementation goes from root to leaf without breaking to ensure everything gets logged, then stores both the revoked and unknown problematic subjects each time it is applicable (since this will ensure leafmost) and choses the correct one to use afterwards. This helps to avoid more complex/less understandable logic that would be needed to do this with iterating from leaf to root or returning early.

Reply:
> Returning early means we wouldn't log every cert

This makes senses.

> If there is a revoked cert closer to the leaf, it wouldn't reflect that cert's subject

We can iterate the loop from the other end ⚡